<!--

Programmer: Victoria Albanese
Filename: camera_right.launch

Purpose: Launch the "right" Intel RealSense 435i camera
on the rigging so that its pointcloud can be visualized

-->

<launch>

    <!-- This launches the camera with the correct parameters -->
    <include file="$(find realsense2_camera)/launch/rs_rgbd.launch">
	<arg name="camera" value="camera_right" />
	<arg name="serial_no" value="843112073834" /> 
    </include>


    <!-- These nodelets limit the FOV of the camera to 0.635 meters in depth -->
    <!-- These nodelets limit the FOV of the camera to (+/-)0.3175 meters to the sides  -->
    <!-- These nodelets filter out the ground plane which is approx. 0.085 meters below the camera-->
    <node pkg="nodelet" type="nodelet" name="cloud_right_filter_fov_depth" args="load pcl/PassThrough pcl_manager" output="screen">
        <remap from="~input"  to="/camera_right/depth_registered/points"/>
        <remap from="~output" to="/camera_right/depth_registered/points_filtered_depth"/>
        <rosparam>
            filter_field_name: z
            filter_limit_min: 0.0
            filter_limit_max: 0.60
            filter_limit_negative: False
        </rosparam>
    </node>	
    <node pkg="nodelet" type="nodelet" name="cloud_right_filter_fov_sides" args="load pcl/PassThrough pcl_manager" output="screen">
        <remap from="~input"  to="/camera_right/depth_registered/points_filtered_depth"/>
        <remap from="~output" to="/camera_right/depth_registered/points_filtered_sides"/>
        <rosparam>
            filter_field_name: x
            filter_limit_min: -0.28
            filter_limit_max:  0.28
            filter_limit_negative: False
        </rosparam>
    </node>	


    <node pkg="nodelet" type="nodelet" name="cloud_right_ground_filter" args="load pcl/PassThrough pcl_manager" output="screen">
        <remap from="~input"  to="/camera_right/depth_registered/points_filtered_sides"/>
        <remap from="~output" to="/camera_right/depth_registered/points_filtered_ground_plane"/>
        <rosparam>
            filter_field_name: z
            filter_limit_min: 0.005
            filter_limit_max: 1.0
            filter_limit_negative: False
        </rosparam>
    </node>	
    <node pkg="nodelet" type="nodelet" name="cloud_right_outlier_removal" args="load pcl/StatisticalOutlierRemoval pcl_manager" output="screen"> 
        <remap from="~input"  to="/camera_right/depth_registered/points_filtered_ground_plane" /> 
        <remap from="~output" to="/camera_right/depth_registered/points_filtered" /> 
        <rosparam> 
               mean_k: 100 
               stddev: 0.1 
               negative: False 
        </rosparam> 
    </node> 


    <!-- This transforms each resulting pointcloud into a global reference frame -->
    <node name="camera_right_tf" pkg="tf" type="static_transform_publisher"
        args="0.317 -0.034 0.098 1.5708 0 -1.60 /world /camera_right_color_optical_frame 100"/>

</launch>

